tokio/src/runtime/basic_scheduler.rs

```diff
@@ -224,156 +217,102 @@ impl<P: Park> BasicScheduler<P> {
-                    if let Ready(v) = crate::coop::budget(|| future.as_mut().poll(&mut cx)) {
-                        return v;
-                    }
-                }
+    fn park(&self, mut core: Box<Core>) -> Box<Core> {
+        let mut park = core.park.take().expect("park missing");
 
-                for _ in 0..MAX_TASKS_PER_TICK {
-                    // Get and increment the current tick
-                    let tick = scheduler.tick;
-                    scheduler.tick = scheduler.tick.wrapping_add(1);
+        if let Some(f) = &self.spawner.shared.before_park {
+            #[allow(clippy::redundant_closure)]
+            let (c, _) = self.enter(core, || f());
+            core = c;
+        }
 
-                    let entry = if tick % REMOTE_FIRST_INTERVAL == 0 {
-                        scheduler.spawner.pop().or_else(|| {
-                            context
-                                .tasks
-                                .borrow_mut()
-                                .queue
-                                .pop_front()
-                                .map(RemoteMsg::Schedule)
-                        })
-                    } else {
-                        context
-                            .tasks
-                            .borrow_mut()
-                            .queue
-                            .pop_front()
-                            .map(RemoteMsg::Schedule)
-                            .or_else(|| scheduler.spawner.pop())
-                    };
+        // This check will fail if `before_park` spawns a task for us to run
+        // instead of parking the thread
+        if core.tasks.is_empty() {
+            // Park until the thread is signaled
+            core.stats.about_to_park();
+            core.stats.submit(&core.spawner.shared.stats);
 
-                    let entry = match entry {
-                        Some(entry) => entry,
-                        None => {
-                            if let Some(f) = &scheduler.before_park {
-                                f();
-                            }
-                            // This check will fail if `before_park` spawns a task for us to run
-                            // instead of parking the thread
-                            if context.tasks.borrow_mut().queue.is_empty() {
-                                // Park until the thread is signaled
-                                scheduler.stats.about_to_park();
-                                scheduler.stats.submit(&scheduler.spawner.shared.stats);
-                                scheduler.park.park().expect("failed to park");
-                                scheduler.stats.returned_from_park();
-                            }
-                            if let Some(f) = &scheduler.after_unpark {
-                                f();
-                            }
+            let (c, _) = self.enter(core, || {
+                park.park().expect("failed to park");
```

Comment:
if this panics do we fail to reset the core back to allow another thread to make progress? Its been a while since I've looked at this code but not sure if the parker panicking also just means the whole runtime has just gone bust.

Reply:
the previous implementation also `expect`ed the attempt to park, fwiw: https://github.com/tokio-rs/tokio/pull/4377/files#diff-084f42aeb27a3c4d2188ca300b539b0ea5808153dde9d6bc3976dae0ba1bf037L282

Reply:
Yeah, but before if I am reading this correctly we placed the scheduler guts back into the option. But here we wouldn't do that? I just remember we went through lengths to protect the panic before so not sure if we covered that here.

Reply:
If park fails, there are bigger issues going on.

Reply:
realistically, I want to remove the `Result` return from park() in general.